File: web/src/types/billing.ts
export interface PlanLimits {
  id: string
  name: string
  uploads_per_day: number
  max_video_seconds: number
  visibility_days?: number | null
  price_display?: string | null
}

export interface UsageSnapshot {
  plan: string
  is_pro: boolean
  uploads_used_today: number
  uploads_remaining_today: number
  uploads_limit: number
  max_video_seconds: number
  visibility_days?: number | null
  period_end: string
  billing_configured: boolean
}

export interface BillingPlansResponse {
  current_plan: string
  is_pro: boolean
  billing_configured: boolean
  plans: PlanLimits[]
  usage: UsageSnapshot
  cancel_at_period_end?: boolean
  subscription_ends_at?: string | null
}

export function videoLimitLabel(seconds: number): string {
  const minutes = Math.max(1, Math.round(seconds / 60))
  return minutes === 1 ? '1 minute' : `${minutes} minutes`
}

export function uploadsLeftLabel(usage: UsageSnapshot): string {
  const left = usage.uploads_remaining_today
  const limit = usage.uploads_limit
  if (left <= 0) return `0 of ${limit} uploads left today`
  return `${left} of ${limit} upload${limit === 1 ? '' : 's'} left today`
}

File: backend/app/schemas/billing.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class PlanLimits(BaseModel):
    id: str
    name: str
    uploads_per_day: int
    max_video_seconds: int
    visibility_days: Optional[int] = None
    price_display: Optional[str] = None


class UsageSnapshot(BaseModel):
    plan: str
    is_pro: bool
    uploads_used_today: int
    uploads_remaining_today: int
    uploads_limit: int
    max_video_seconds: int
    visibility_days: Optional[int] = None
    period_end: str
    billing_configured: bool = False


class BillingPlansResponse(BaseModel):
    current_plan: str
    is_pro: bool
    billing_configured: bool
    plans: list[PlanLimits]
    usage: UsageSnapshot
    cancel_at_period_end: bool = False
    subscription_ends_at: Optional[str] = None


class CheckUploadRequest(BaseModel):
    video_duration: Optional[float] = Field(default=None, ge=0)


class CheckoutRequest(BaseModel):
    success_path: str = "/billing/success"
    cancel_path: str = "/profile"


class CheckoutResponse(BaseModel):
    url: str


class PortalResponse(BaseModel):
    url: str
